src/compas/files/base_reader.py

```diff
@@ -0,0 +1,188 @@
+from __future__ import absolute_import
+from __future__ import division
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+try:
+    from urllib.request import urlretrieve
+except ImportError:
+    from urllib import urlretrieve
+
+
+class BaseReader(object):
+    """Base class containing file reading functions for file extension specific readers
+
+    Attributes
+    ----------
+    location : str or pathlib object
+        Path or URL to the file
+
+    """
+    def __init__(self, location):
+        self._location = location
+        self._is_binary = None
+
+    @property
+    def location(self):
+        """ Create or ensure Pathlib object for file
+
+        Parameters
+        ----------
+        location : string or Pathlib object
+            location specified either as an URL, string containing path or
+            Pathlib object
+
+        Raises
+        ------
+        IOError
+            If file is not found
+        """
+        if self.is_location_url():
+            pathobj = self.download(self._location)
+        else:
+            if not isinstance(self._location, Path):
+                pathobj = Path(self._location)
+            else:
+                pathobj = self._location
+
+        if pathobj.exists():
+            return pathobj
+
+        raise IOError('File not found.')
+
+    @property
+    def is_binary(self):
+        pass
+
+    @property
+    def is_valid(self):
+        return NotImplementedError
+
+    def is_location_url(self):
+        """Checks if given location is a string containing an URL
+
+        Returns
+        -------
+        bool
+            True if recognized as an URL
+        """
+        self._is_url = str(self._location).startswith('http')
+        return self._is_url
```

Comment:
i think this should be 
```python
return str(self._location).startswith('http')
```
that said, i think it is a bad idea to use `self.location` and `self._location` in this way, because it is something you need to know to do it correctly.
perhaps it would be better to differentiate between the original address provided by the user, which could be a url, a relative path, an absolute path, and a `Path` object. the reader processes this into a new variable that always refers to an absolute local file path in the form of a `Path` object. assuming the original address provided by the user is stored as `self._address` this function would just do
```python
return str(self._address).startswith('http')
```